node/common/client_service_base.ts

```diff
@@ -291,10 +301,257 @@ export class ClientServiceBase {
       .build();
   }
 
+  /**
+   * Execute a registered contract
+   * @deprecated Use {@link execute} instead
+   * @param {string} contractId
+   * @param {Object} contractArgument
+   * @param {Object|null} [functionArgument=null]
+   * @return {Promise<ContractExecutionResult|void|*>}
+   * @throws {ClientError|Error}
+   */
+  async executeContract(
+    contractId: string,
+    contractArgument: Object,
+    functionArgument: Object | null = null
+  ): Promise<ContractExecutionResult> {
+    return this.execute(contractId, contractArgument, null, functionArgument);
+  }
+
+  /**
+   * Execute a registered contract and function (optionally)
+   * @param {string} contractId
+   * @param {Object|string} contractArgument
+   * @param {string|null} [functionId=null]
+   * @param {Object|string|null} [functionArgument=null]
+   * @param {string|null} [nonce=null]
+   * @return {Promise<ContractExecutionResult>}
+   * @throws {ClientError|Error}
+   */
+  async execute(
+    contractId: string,
+    contractArgument: Object | string,
+    functionId: string | null = null,
+    functionArgument: Object | string | null = null,
+    nonce: string | null = null
```

Comment:
It might be a stupid question, but this kind of notation can accept function calls that omit some arguments (e.g., `execute("contract", "...")` )? Don't need `?`? Sorry, I'm a super beginner at TS.

Reply:
@jnmt 
Thank you for your question.
It's a good question in my opinion.

I am not 100% sure if my understanding ↓ is correct.

If the optional argument has an explicit default value, then we don't need the `?` notation.

In the case of

```
func(a: string, b?: string)
```

the b's type definition equals `string | undefined` (because an unassigned optional argument is initiated with `undefined`).

To me, it's a sugar of
```
func(a: string, b: string | undefined = undefined)
```

This is my understanding 😄

Reply:
Thank you for the detailed explanation. Interesting. I understand `string | null = null` is an explicit notation (and the only way) to accept the omission and initiate it with `null`.